06_HelloGraphicsQueue/main.cpp

```diff
@@ -2,80 +2,436 @@
+		IGPUCommandBuffer::SImageMemoryBarrier res;
+		res.barrier.srcAccessMask = E_ACCESS_FLAGS::EAF_NONE;
+		res.barrier.dstAccessMask = E_ACCESS_FLAGS::EAF_NONE;
+		res.oldLayout = asset::IImage::EL_UNDEFINED;
+		res.newLayout = asset::IImage::EL_UNDEFINED;
+		res.srcQueueFamilyIndex = 0u;
+		res.dstQueueFamilyIndex = 0u;
+		res.image = nullptr;
+		res.subresourceRange = constructDefaultInitializedSubresourceRange();
+
+		return res;
+	}
+
+	IGPUImage::SCreationParams constructImageCreationParams(const VkExtent2D& extent) const
+	{
+		IGPUImage::SCreationParams res{};
+		res.type = IImage::E_TYPE::ET_2D;
+		res.extent.height = 512u;
+		res.extent.width = 512u;
+		res.extent.depth = 1u;
+		res.format = asset::E_FORMAT::EF_R8G8B8A8_SRGB;
+		res.mipLevels = 1u;
+		res.flags = IImage::ECF_NONE;
+		res.arrayLayers = 1u;
+		res.samples = IImage::E_SAMPLE_COUNT_FLAGS::ESCF_1_BIT;
+		res.tiling = video::IGPUImage::ET_OPTIMAL;
+		res.usage = asset::IImage::EUF_TRANSFER_SRC_BIT | asset::IImage::EUF_TRANSFER_DST_BIT;
+		res.queueFamilyIndexCount = 0u;
+		res.queueFamilyIndices = nullptr;
+		res.initialLayout = asset::IImage::EL_UNDEFINED;
+
+		return res;
+	}
+
+	core::smart_refctd_ptr<IGPUImage> createImage(const VkExtent2D& extent)
+	{
+		IGPUImage::SCreationParams imgParams{};
+		imgParams.type = IImage::E_TYPE::ET_2D;
+		imgParams.extent.height = extent.height;
+		imgParams.extent.width = extent.width;
+		imgParams.extent.depth = 1u;
+		imgParams.format = asset::E_FORMAT::EF_R8G8B8A8_SRGB;
+		imgParams.mipLevels = 1u;
+		imgParams.flags = IImage::ECF_NONE;
+		imgParams.arrayLayers = 1u;
+		imgParams.samples = IImage::E_SAMPLE_COUNT_FLAGS::ESCF_1_BIT;
+		imgParams.tiling = video::IGPUImage::ET_OPTIMAL;
+		imgParams.usage = asset::IImage::EUF_TRANSFER_SRC_BIT | asset::IImage::EUF_TRANSFER_DST_BIT;
+		imgParams.queueFamilyIndexCount = 0u;
+		imgParams.queueFamilyIndices = nullptr;
+		imgParams.initialLayout = asset::IImage::EL_UNDEFINED;
+
+		auto img = m_device->createImage(std::move(imgParams));
+
+		const size_t imageByteSize = img->getImageDataSizeInBytes();
+		nbl::video::IGPUBuffer::SCreationParams params = {};
+		params.size = imageByteSize;
+		params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT;
+		smart_refctd_ptr<IGPUBuffer> imageBuffer = m_device->createBuffer(std::move(params));
+		if (!imageBuffer)
 		{
-			return {{.requiredFlags=flags_t::EQF_GRAPHICS_BIT,.disallowedFlags=flags_t::EQF_NONE,.queueCount=1,.maxImageTransferGranularity={1,1,1}}};
+			logFail("Failed to create a GPU Buffer of size %d!\n", params.size);
+			return nullptr;
 		}
+		auto allocation = m_device->allocate(imageBuffer->getMemoryReqs(), imageBuffer.get(), nbl::video::IDeviceMemoryAllocation::EMAF_NONE);
+
+		ILogicalDevice::SBindImageMemoryInfo bindMemInfo;
+		bindMemInfo.image = img.get();
+		bindMemInfo.memory = allocation.memory.get();
+		bindMemInfo.offset = 0ull;
+
+		m_device->bindImageMemory(1u, &bindMemInfo);
```

Comment:
make sure you use dedicated allocations for the image

Reply:
In general on Windows, you get a hard limit of 4096 vkDeviceMemory allocations.

Which is why you're supposed to combine and suballocate multiple resources from same memory as long as your heap requirements match (type and attributes).

However, because:
- some non-Vulkan APIs don't have a concept of separable resource and memory allocation binding (i.e. when we used GL in the past) or
- you're importing/exporting images and buffers to/from DX12, CUDA, etc., or
- there's a performance improvement (i.e. Nvidia framebuffer images, or self-created swapchain images)
you might have to use dedicated allocations (only 1 resource binds to a particular vkDevicememory)

In Nabla we do it by passing the resource the allocation will be dedicated to as part of the memory allocation method:
https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/blob/aff738da77804a39684c3053867d77b03c5fd803/02_HelloCompute/main.cpp#L195

Here you requested a dedication, which means the memory is ALREADY bound, the latter `m_device->bindImageMemory` won't actually succeed and you'll get an error, its not needed.